pkg/api/helpers/download.go

```diff
@@ -245,48 +235,107 @@ func (d *Downloader) downloadPresignedPart(ctx context.Context, physicalAddress
 		buf = buf[:partSize]
 	}
 
-	_, err = io.ReadFull(resp.Body, buf)
+	var err error
+	for attempt := 0; attempt <= d.BodyRetries; attempt++ {
+		if attempt > 0 {
+			// sleep for a random time between 0 and 1 second or break on context cancellation
+			select {
+			//nolint:gosec
+			case <-time.After(time.Duration(rand.IntN(DefaultDownloadRetryDelayMs)) * time.Millisecond):
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+		}
+
+		// create request with range header
+		var req *http.Request
+		req, err = http.NewRequestWithContext(ctx, http.MethodGet, physicalAddress, nil)
+		if err != nil {
+			return err
+		}
+		req.Header.Set("Range", rangeHeader)
+
+		err = executeHTTPRequest(d.HTTPClient, req, func(resp *http.Response) error {
+			if resp.StatusCode != http.StatusPartialContent {
+				return fmt.Errorf("%w: %s", ErrRequestFailed, resp.Status)
+			}
+			if resp.ContentLength != partSize {
+				return fmt.Errorf("%w: part %d expected %d bytes, got %d", ErrRequestFailed, partNumber, partSize, resp.ContentLength)
+			}
+			_, readErr := io.ReadFull(resp.Body, buf)
+			return readErr
+		})
+		if err == nil {
+			break
+		}
+	}
 	if err != nil {
-		return err
+		return fmt.Errorf("failed to download part %d after %d retries: %w", partNumber, d.BodyRetries+1, err)
 	}
 
 	_, err = f.WriteAt(buf, rangeStart)
-	if err != nil {
-		return err
-	}
-	return nil
+	return err
 }
 
 func (d *Downloader) downloadObject(ctx context.Context, src uri.URI, dst string, tracker *progress.Tracker) error {
-	// get object content
-	resp, err := d.Client.GetObject(ctx, src.Repository, src.Ref, &apigen.GetObjectParams{
-		Path:    *src.Path,
-		Presign: swag.Bool(d.PreSign),
-	})
-	if err != nil {
-		return err
-	}
-	defer func() { _ = resp.Body.Close() }()
-	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("%w: %s", ErrRequestFailed, resp.Status)
-	}
+	var err error
+	for attempt := 0; attempt <= d.BodyRetries; attempt++ {
+		if attempt > 0 {
+			// sleep for a random time between 0 and 1 second or break on context cancellation
+			select {
+			//nolint:gosec
+			case <-time.After(time.Duration(rand.IntN(DefaultDownloadRetryDelayMs)) * time.Millisecond):
+			case <-ctx.Done():
+				return ctx.Err()
+			}
+			// Remove destination file if retrying (will be recreated)
+			_ = os.Remove(dst)
```

Comment:
This is hard to maintain, since it's so easy to miss this when the code below evolves. 
WDYT about truncating the file upon first write instead?